lib/core/src/Cardano/Wallet/Primitive/Types/TokenQuantity.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Wallet.Primitive.Types.TokenQuantity
+    (
+      -- * Type
+      TokenQuantity (..)
+
+      -- * Values
+    , zero
+
+      -- * Operations
+    , add
+    , subtract
+    , negate
+    , pred
+    , succ
+
+      -- * Tests
+    , isStrictlyPositive
+    , isStrictlyNegative
+    , isNonZero
+    , isZero
+
+    ) where
+
+import Prelude hiding
+    ( negate, pred, subtract, succ )
+
+import Control.DeepSeq
+    ( NFData (..) )
+import Data.Aeson
+    ( FromJSON (..), ToJSON (..) )
+import Data.Text.Class
+    ( FromText (..), ToText (..) )
+import Fmt
+    ( Buildable (..) )
+import GHC.Generics
+    ( Generic )
+import Quiet
+    ( Quiet (..) )
+
+import qualified Prelude
+
+--------------------------------------------------------------------------------
+-- Type
+--------------------------------------------------------------------------------
+
+newtype TokenQuantity = TokenQuantity
+    { unTokenQuantity :: Integer }
+    deriving stock (Eq, Ord, Generic)
+    deriving (Read, Show) via (Quiet TokenQuantity)
+
+--------------------------------------------------------------------------------
+-- Instances
+--------------------------------------------------------------------------------
+
+instance NFData TokenQuantity
+
+instance Semigroup TokenQuantity where
+    (<>) = add
+
+instance Monoid TokenQuantity where
+    mempty = zero
+
+instance Buildable TokenQuantity where
+    build = build . toText . unTokenQuantity
+
+instance ToText TokenQuantity where
+    toText = toText . unTokenQuantity
+
+instance FromText TokenQuantity where
+    fromText = fmap (TokenQuantity . fromIntegral @Integer) . fromText
+
+instance FromJSON TokenQuantity where
+    parseJSON = fmap TokenQuantity . parseJSON
+instance ToJSON TokenQuantity where
+    toJSON = toJSON . unTokenQuantity
+
+--------------------------------------------------------------------------------
+-- Values
+--------------------------------------------------------------------------------
+
+zero :: TokenQuantity
+zero = TokenQuantity 0
+
+--------------------------------------------------------------------------------
+-- Operations
+--------------------------------------------------------------------------------
+
+add :: TokenQuantity -> TokenQuantity -> TokenQuantity
+add (TokenQuantity x) (TokenQuantity y) = TokenQuantity $ x + y
+
+subtract :: TokenQuantity -> TokenQuantity -> TokenQuantity
```

Comment:
Is subtract used anywhere?
And what happens if its result is negative?
I think this function should either be failable (i.e. Maybe TokenQuantity) or removed.

Reply:
@rvl I've now removed this function.

I think you're right that if we need to compare or find the difference between two quantities, there are other, safer methods that we can use. 👍🏻 

Fixed in 9a5bc7ea71dd4e7ab46b4e02f6b04d12badd0686.

Reply:
Thanks.